dolphin/drivers/manager.py

```diff
@@ -34,63 +35,59 @@ def __init__(self):
         # and storage system is effectively used.
         self.driver_factory = dict()
 
-    @staticmethod
-    def get_storage_registry():
-        """Show register parameters which the driver needs."""
-        pass
+    def _create_driver(self, context, **kwargs):
+        vendor = kwargs.get('vendor', None)
+        model = kwargs.get('model', None)
 
-    def register_storage(self, context, access_info):
-        """Discovery a storage system with access information."""
         try:
+            if not vendor or not model:
+                access_info = helper.get_access_info(context, kwargs['storage_id'])
+                kwargs = copy.deepcopy(access_info)
+
             driver = stevedore.driver.DriverManager(
                 namespace=self.NAMESPACE,
-                name='%s %s' % (access_info['vendor'],
-                                access_info['model']),
-                invoke_on_load=True
+                name='%s %s' % (kwargs['vendor'], kwargs['model']),
+                invoke_on_load=True,
+                invoke_kwds=kwargs
             ).driver
         except Exception as e:
-            msg = (_("Storage driver '%s %s' could not be found.") % (access_info['vendor'],
-                                                                      access_info['model']))
-            LOG.error(msg)
+            msg = (_("Storage driver '%s %s' could not be found.") % (kwargs['vendor'],
+                                                                      kwargs['model']))
+            LOG.error(e)
             raise exception.StorageDriverNotFound(message=msg)
 
-        storage = driver.register_storage(context,
-                                          access_info)
-        if storage:
-            storage['id'] = six.text_type(uuidutils.generate_uuid())
-            self.driver_factory[storage['id']] = driver
-
-        LOG.info("Storage was found successfully.")
-        return storage
-
-    def remove_storage(self, context, storage_id):
+        return driver
+
+    def get_driver(self, context, **kwargs):
+        """
+        Caller can get driver with storage_id or vendor&model
+        :param context:
+        :param kwargs: A dictionary, only storage_id is necessary, for the first
+            time to register storage, storage_id is newly created, nothing can be found
+            from database with this storage_id, so other access information should be
+            included in access_info, otherwise, storage_id can be used to get access
+            information from database.
+        :return: A driver object.
+        """
+        storage_id = kwargs.get('storage_id', None)
+        if not storage_id:
+            raise exception.StorageDriverNotFound(message="storage_id is missed")
```

Comment:
it's not suitable use StorageDriverNotFound here, suggest use InvalidParameter or others

Reply:
Refactored, do not check storage_id here anymore.